File: src/imu_pkg/src/imu_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <tf/tf.h>
#include <geometry_msgs/Twist.h>
ros::Publisher pub_vel;
void imu_cb(sensor_msgs::ImuConstPtr imu)
{
    if (imu->orientation_covariance[0] < 0)
    {
        return;
    }
    tf::Quaternion quaternion(
        imu->orientation.x,
        imu->orientation.y,
        imu->orientation.z,
        imu->orientation.w);
    double roll, pitch, yaw;
    tf::Matrix3x3(quaternion).getRPY(roll, pitch, yaw);
    ROS_INFO("翻滚：%.2f,俯仰：%.2f,偏航：%.2f,", roll, pitch, yaw);
    geometry_msgs::Twist cmd_vel;
    double target_yaw = 3.1415926;
    double diff_angle = target_yaw - yaw;
    cmd_vel.angular.z = diff_angle * 0.5;
    pub_vel.publish(cmd_vel);
}
int main(int argc, char *argv[])
{
    setlocale(LC_ALL, "");
    ros::init(argc, argv, "imu_node");
    ros::NodeHandle nh;
    ros::Subscriber imu_sub = nh.subscribe("imu/data", 1, imu_cb);
    pub_vel = nh.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
    ros::spin();
    return 0;
}

File: src/imu_pkg/scripts/imu_node.py
#!/usr/bin/env python3
import rospy
from sensor_msgs.msg import Imu
from tf.transformations import euler_from_quaternion
from geometry_msgs.msg import Twist
pub_vel = rospy.Publisher("cmd_vel",Twist,queue_size=1)
def imu_cb(imu=Imu()):
    if imu.orientation_covariance[0]<0:
        return
    quaternion = [ 
        imu.orientation.x,
        imu.orientation.y,
        imu.orientation.z,
        imu.orientation.w]
    roll,pitch,yaw = euler_from_quaternion(quaternion)
    rospy.loginfo("滚转:%.2f,俯仰:%.2f,偏航:%.2f",roll,pitch,yaw)
    target_yaw = -1;
    diff_angle = target_yaw - yaw
    cmd_vel = Twist()
    cmd_vel.angular.z = diff_angle*0.5
    pub_vel.publish(cmd_vel)
    
def main():
    rospy.init_node("imu_node_py")
    sub_imu = rospy.Subscriber("imu/data",Imu,imu_cb,queue_size=1)
    rospy.spin()
if __name__ == '__main__':
    main()
    